### PR TITLE
[Lotto - Step 1] Submit miooo0o's mission

### PR DESCRIPTION
**Pull request by miooo0o**
This pull request adds a console-based Lotto program.  
It includes the core game logic, basic error handling, and a structure that handles retries based on user input.

---
### TDD Experience

When I did the pre-course, I tried practicing TDD, but I was too focused on “doing it perfectly,” so I didn’t get to practice it properly, which I regretted. This time, I started with higher expectations. After watching Jason’s live coding session, I got a clearer idea of TDD, and Sebas and I first talked about how we wanted to approach the problem.  

I wanted to stick to TDD step by step, even if it took more time, and I thought it was meaningful to really experience TDD properly, even if we couldn’t finish everything in one day. On the other hand, Sebas thought we needed at least a working program to get feedback, so he wanted to prioritize having functional code, even if we had to skip parts of TDD.

So, we agreed to follow these points:

- Write a feature list first if possible.
- Don’t get too obsessed with writing “perfect” code.
- Try to follow TDD as much as we can.
- Skip parts that both of us agree take too much time to test.
- Write as many unit tests as possible.
- Make sure to leave enough time for refactoring later.
    
As a result, we finished a working version on July 2, 2025, and started refactoring the next day.

In addition to the code itself, I would really appreciate feedback on these points:

1. **About the order and number of commits**
    - We tried to split commits into small pieces and reflect changes whenever needed. I’m still working on building good commit habits, so I’d appreciate your thoughts on whether our approach was reasonable or if there’s anything we could improve.
        
2. **About our coding style and understanding of the testing/TDD flow**
    - We didn’t manage to write an E2E test, and although we were supposed to write the input/output view using TDD, we prioritized writing the classes first because the tests would have taken too long. We followed TDD as much as possible but skipped some parts when necessary, and we split test writing and method/class implementation when needed. I’d appreciate any advice on how to approach this better next time.
        
3. **Whether the code quality and structure were appropriate for the project’s size**
    - For example, we didn’t separate the Service part because the application is still small and simple, but during refactoring, we felt it looked a bit cluttered. Should we have separated responsibilities more, even for a project of this size?
    - Also, since English is not my native language, I tried to make method and class names as clear as possible during the refactor. I’d like to know if you think our naming was appropriate.
        

Thank you so much for reading this long message. I hope you have a great day!

---

### TDD 경험

프리 코스 때 TDD를 시도해봤지만, ‘잘 해내야 한다’는 부담이 너무 커서 제대로 연습하지 못한 게 아쉬웠습니다. 그래서 이번에는 더 큰 기대를 가지고 시작했습니다. Jason의 라이브 코딩을 보고 나니 TDD에 대한 감이 조금 더 잡혔고, Sebas와 함께 어떻게 진행할지 먼저 이야기 나눴습니다.  
저는 시간이 좀 더 걸리더라도 단계마다 TDD를 지켜보고 싶었고, 오늘 안에 완성이 안 되더라도 TDD를 충분히 경험하는 데 의미를 두고 싶었습니다. 반면 Sebas는 최소한 동작하는 코드를 먼저 만들어야 피드백을 받을 수 있으니, 필요하다면 TDD를 일부 생략하자는 의견이었습니다.

그래서 우리는 이렇게 정리하고 진행하기로 했습니다:

- 가능하다면 먼저 기능 목록(feature list)을 작성한다.
- 완벽한 코드에 집착하지 않는다.
- 할 수 있는 만큼 TDD를 시도한다.
- 둘 다 동의하거나, 테스트에 너무 많은 시간이 걸리는 부분은 과감히 생략한다.
- 최대한 모듈 단위로 테스트한다.
- 이후에 꼭 리팩터링할 시간을 남긴다.

그 결과, 2025년 7월 2일에 동작하는 코드를 완성했고, 다음날 refactor를 시작했습니다.

추가로 코드 외에 꼭 받고 싶은 피드백은 아래와 같습니다:

1. **커밋의 순서와 횟수에 대해**
    - 커밋을 최대한 잘게 나누고, 변경사항이 있을 때마다 반영하려고 노력했습니다. 아직 좋은 커밋 습관을 만드는 중인데, 이번 방식이 괜찮았는지, 더 보완할 점이 있다면 알려주시면 감사하겠습니다.

2. **코딩 방식과 테스트/TDD 흐름에 대해**
    - E2E 테스트는 나중에 만들기로 했는데, 만들지는 못했습니다. 또 TDD로 input/output view까지 작성해야 했지만 테스트가 너무 길어질 것 같아 우선 클래스를 먼저 작성했습니다. TDD를 따르되 필요한 부분은 생략하고, 테스트 작성과 메서드/클래스 작성은 나눠서 진행하는 등 나름의 타협점을 찾았습니다. 다음에는 이런 상황에서 어떻게 접근하면 좋을지 조언 부탁드립니다.
        
3. **코드의 품질과 구조가 프로젝트 규모에 적절했는지에 대해**
    - 특히 Service 부분은 프로그램이 아직 작고 단순해서 따로 분리하지 않았지만, 리팩터링할 때 약간 산만하다고 느꼈습니다. 이 정도 규모에서도 책임을 더 나눠야 했을지 궁금합니다.
    - 이번 리팩터링에서는 메서드나 클래스 이름이 최대한 명확해지도록 신경 썼는데, 네이밍이 적절했는지도 피드백 주시면 감사하겠습니다.
    

긴 글 읽어주셔서 감사합니다. 좋은 하루 보내세요!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,63 @@
...
+### Controller
+
+- [x] Add the controller where happen the flow of the app
+- [x] Use View classes for input and output
+- [x] Create a retry help functions to validate de input and in case of error retry it.
+- [x] Create a method or service to create the Lotto tickets
+- [x] Create retry helper function to get user input
+- [x] Add a generateLottoTickets service
+
+### Services
```
Comment (Hyeon9mak) [inline README.md:30]: Nice!

Organizing business or feature requirements in documents like the README.md (or something), will be helpful in the future whether it's you revisiting the project later on, or a fellow developer joining the team and looking at the project for the first time.

and it seems to match perfectly with git commit. nice work! 👍 

Hope you keep this habit going and keep refining it!

File: README.md [hunk 2]
```diff
@@ -1 +1,63 @@
...
+- [x] customize exception for DRY
+- [x] add documentation
+- [x] separate controller logics
+- [x] refactor methods
+
+---
+
+### TODOS
+
+- [ ] add **E2E test**
```
Comment (Hyeon9mak) [inline README.md]: That would be nice 😉 👍

File: src/main/kotlin/lotto/Application.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ */
+fun main() {
+    try {
+        val controller = LottoController()
+        controller.run()
+    } catch (e: Exception) {
+        println(e.message ?: "Unexpected error occurred.")
+        exitProcess(1)
+    }
+}
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/Application.kt]: when switching from console screen to web browser, it seems like both the `view` package and `Application.kt` need to change!
is it also the View's responsibility to decide whether to pass error information as strings or as images? 😄

File: src/main/kotlin/lotto/controller/LottoController.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+import lotto.domain.Lotto
+import lotto.domain.Rank
+import lotto.domain.WinningCombination
+import lotto.services.LottoService
+import lotto.util.LottoUtils
+import lotto.view.InputView
+import lotto.view.OutputView
+
+class LottoController {
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/controller/LottoController.kt]: `LottoController` also has no properties like View classes.
how about it makes sense to turn it into an object class?
what are the pros & corns of doing that?

File: src/main/kotlin/lotto/util/LottoUtils.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        }
+    }
+
+    /**
+     * @return true if the number is in the range 1 to 45
+     */
+    fun isInRange(number: Int): Boolean {
+        return number in Lotto.MIN_RANGE..Lotto.MAX_RANGE
+    }
+}
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/util/LottoUtils.kt]: The class named `LottoUtils`, we’ll probably end up opening this file every time we need to add, modify, or review any lotto functionality.
Over time, it could easily grow into a massive file with thousands of lines of code.

I think we need try to avoid such generic names whenever possible and prefer to split things up with more descriptive and specific names.

What do you think about this?

File: src/main/kotlin/lotto/util/LottoUtils.kt [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+            throw exception
+        }
+    }
+
+    /**
+     * @return true if the number is in the range 1 to 45
+     */
+    fun isInRange(number: Int): Boolean {
+        return number in Lotto.MIN_RANGE..Lotto.MAX_RANGE
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/util/LottoUtils.kt]: Let’s say a request comes in to change the "lotto numbers to a range from 1 to 50".
Which class do you think another developer would open first — `Lotto` or `LottoUtil`?
I believe thinking about things like this — making the code easier to read and navigate for others in the future — is what OOP is really about.

File: src/main/kotlin/lotto/controller/LottoController.kt [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            val result = LottoService.getRankFrom(ticket, winningCombination)
+            ranks.add(result)
+        }
+        return ranks
+    }
+
+    private fun getTickets(): Pair<Int, List<Lotto>> {
+        val amount =
+            LottoUtils.retryUntilSuccess({
+                InputView.getPurchaseAmount().also { LottoService.purchaseAmountValidator(it) }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/controller/LottoController.kt]: Wouldn't the readability improve if we arranged the functions according to the flow of execution?
My thought is `readability == consistency == productivity` 😄

File: src/main/kotlin/lotto/domain/Lotto.kt [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+class Lotto(val numbers: List<Int> = emptyList<Int>()) {
+    init {
+        LottoUtils.requireOrThrow(
+            numbers.size == LOTTO_PICK_SIZE,
+            LottoException.InvalidWinningNumbersException(),
+        )
+        LottoUtils.requireOrThrow(
+            numbers.distinct().size == LOTTO_PICK_SIZE,
+            LottoException.InvalidWinningNumbersException(),
+        )
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/domain/Lotto.kt]: Since duplicates aren’t allowed, how about using data structure better suited than a List? 😉

File: src/main/kotlin/lotto/domain/Lotto.kt [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        LottoUtils.requireOrThrow(
+            numbers.all { LottoUtils.isInRange(it) },
+            LottoException.InvalidWinningNumbersException(),
+        )
+    }
+
+    override fun toString(): String {
+        return numbers.joinToString(", ", "[", "]")
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/domain/Lotto.kt]: And also when switching from console screen to web browser, it seems like both the view package and this!
how do you think?

File: src/main/kotlin/lotto/domain/Rank.kt [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+            matchBonus: Boolean = false,
+        ): Rank {
+            if (countOfMatch == SECOND.countOfMatch) {
+                return if (matchBonus) SECOND else THIRD
+            }
+            return entries.find {
+                it.countOfMatch == countOfMatch
+            } ?: MISS
+        }
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/domain/Rank.kt:27]: in the future, if winningMoney changes or new rank are added, only need to look at the Rank enum class! 💯 👍

File: src/main/kotlin/lotto/domain/WinningCombination.kt [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    init {
+        LottoUtils.requireOrThrow(
+            LottoUtils.isInRange(bonusNumber),
+            LottoException.InvalidBonusNumberException(bonusNumber),
+        )
+        LottoUtils.requireOrThrow(
+            winningTicket.numbers.none { it == bonusNumber },
+            LottoException.InvalidBonusNumberException(bonusNumber),
+        )
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/domain/WinningCombination.kt]: I think the `Lotto`'s number and `bonusNumber` are same thing.
how about wrapping both of them with the same VO (Value Object), and moving the validation there?
it will removes a lot of same codes!

File: src/main/kotlin/lotto/domain/WinningCombination.kt [hunk 12]
```diff
@@ -0,0 +1,20 @@
+package lotto.domain
+
+import lotto.exceptions.LottoException
+import lotto.util.LottoUtils
+
+/**
+ * the class contain winning ticket and bonus number to eval user's Lotto tickets
+ */
+class WinningCombination(val winningTicket: Lotto, val bonusNumber: Int = 0) {
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/domain/WinningCombination.kt]: What reason for `bonusNumber`'s default value `0`?
if there is no reason, then plz remove that default value.
it makes confuse of another developer.

File: src/main/kotlin/lotto/services/LottoService.kt [hunk 13]
```diff
@@ -0,0 +1,51 @@
+package lotto.services
+
+import lotto.domain.Lotto
+import lotto.domain.Rank
+import lotto.domain.WinningCombination
+import lotto.exceptions.LottoException
+
+object LottoService {
+    private const val ZERO_DOUBLE = 0.0
+
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/services/LottoService.kt]: why it called `Service`?
and why need this class?

it just curious about your thoughts. plz feel free to answer. 😄

File: src/main/kotlin/lotto/services/LottoService.kt [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    fun purchaseAmountValidator(amount: Int) {
+        if (amount < Lotto.Companion.PRICE_OF_TICKET || amount % Lotto.Companion.PRICE_OF_TICKET != 0) {
+            throw LottoException.InvalidAmountException(
+                amount,
+            )
+        }
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/services/LottoService.kt]: How about ask the validation to object directly?

```kotlin
Lotto.isValidatePurchase(...) // or something
Lotto(purchase = purchase) // validation logic in constructor
```

And what are the advantages of this method(or consructor)?

File: src/main/kotlin/lotto/view/InputView.kt [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    fun getWinningNumbers(): List<Int> {
+        println(Constants.WINNING_NUMBERS_PROMPT)
+        val input = readln()
+        return input
+            .split(',')
+            .map {
+                it.trim().toIntOrNull()
+                    ?: throw LottoException.InvalidWinningNumbersFormatException(input)
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/view/InputView.kt]: yep! the View's responsibility to determine how to parse the string. 😄 👍

File: src/main/kotlin/lotto/view/InputView.kt [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+        println(Constants.BONUS_NUMBER_PROMPT)
+        val input = readln()
+        return input.trim().toIntOrNull() ?: throw LottoException.InvalidBonusNumberFormatException(input)
+    }
+
+    object Constants {
+        const val PURCHASE_PROMPT = "Please enter the purchase amount."
+        const val WINNING_NUMBERS_PROMPT = "Please enter last week’s winning numbers."
+        const val BONUS_NUMBER_PROMPT = "Please enter the bonus number."
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/view/InputView.kt]: Kotlin object class is a single tone support class,
so it's okay to declare a constant without a inner object class.

```kotlin
object InputView {
    private const val PURCHASE_PROMPT = "Please enter the purchase amount."
    private const val WINNING_NUMBERS_PROMPT = "Please enter last week’s winning numbers."
    private const val BONUS_NUMBER_PROMPT = "Please enter the bonus number."

    // ...
}
```

File: src/main/kotlin/lotto/view/OutputView.kt [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+        val matchCount = results.count { it.countOfMatch == entry.countOfMatch }
+        val hasBonus = if (entry.requiresBonus) " + Bonus Ball" else ""
+        val winningMoney = "%,d".format(entry.winningMoney)
+        val pluralizedTicket = pluralizeTicket(matchCount)
+        println("${entry.countOfMatch} Matches$hasBonus ($winningMoney ${Lotto.CURRENCY}) - $matchCount $pluralizedTicket")
+    }
+
+    private fun pluralizeTicket(size: Int): String {
+        return if (size == 1) "ticket" else "tickets"
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/view/OutputView.kt]: detail! 😮 👍

File: src/test/kotlin/lotto/domain/LottoTest.kt [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+        assertThat(Lotto(ticket))
+    }
+
+    @Test
+    fun `Lotto numbers are not in range`() {
+        val ticket = listOf(1, 100, 2, 3, 4, 5)
+        assertThrows<IllegalArgumentException> {
+            Lotto(ticket)
+        }
+    }
```
Comment (Hyeon9mak) [inline src/test/kotlin/lotto/domain/LottoTest.kt]: many test case 👍 💯 
How about also testing with boundary values?

listOf(1, 2, 3, 4, 5, 46)
listOf(0, 2, 3, 4, 5, 6)
listOf(1, 2, 3, 4, 5)
listOf(1, 2, 3, 4, 5, 6, 7)

File: src/main/kotlin/lotto/services/LottoService.kt [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+        return Rank.valueOf(countOfMatch, matchBonus)
+    }
+
+    fun calculateReturnRate(
+        totalAmount: Int,
+        results: List<Rank>,
+    ): Double {
+        val totalPrize = results.sumOf { it.winningMoney }
+        return if (totalPrize == 0) ZERO_DOUBLE else (totalPrize.toDouble() / totalAmount) * 100.0
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/services/LottoService.kt]: as you know, the `LottoService` class have a lot of logic.
it means every request change from business, will open this class and need changed.
that makes lot of merge conflict & your headache.

i think you can make domain class for each methods!

**Review by Hyeon9mak — CHANGES_REQUESTED** [review]
Hello Mina!
I've tried my best to leave thorough reviews.
Below are the answers to your questions.

### About the order and number of commits
Considering the origin of Git, Git commits are clear turning points that can be easily reversed when problems arise.
Smaller commits with detailed messages make this process easier.

Also, every commit must pass the build.
That way, there will be no issues for fellow developers when returning to that commit.

I think you did a good job with it!

### About our coding style and understanding of the testing/TDD flow
Did you really need I/O view test and E2E test?

The most important reason for TestCode is to clear the "unstable parts" of the program before you run it yourself.
I think it's enough to focus on a domain classes - that contains a lot of really complicated and difficult logic.

And I think it's a good strategy to boldly omit the parts that you think are unnecessary. Wouldn't it be more precious to have another coffee with Sebas at that time? 😄 

### Whether the code quality and structure were appropriate for the project’s size

left review comments on this, so please check them out.

Feel free to ask any questions you have!

File: src/main/kotlin/lotto/Application.kt (hunk 3 above)
Comment (miooo0o) [reply src/main/kotlin/lotto/Application.kt]: Honestly, I haven’t really thought about exactly how much `Application.kt` should ideally handle.  
I was just thinking of it like doing what `int main(int ac, char** av)` would do in languages I used before.

Is it more appropriate for the `view` package to handle something like this?

```kotlin
println(e.message ?: "Unexpected error occurred.")
```

For extra safety, we agreed to add another `try/catch` so that if anything other than `IllegalArgumentException` is thrown, it just terminates.
I did think about passing it to `OutputView` instead, but we didn’t really reach a clear conclusion about whether we should keep exposing `println()` directly for message handling — so the discussion was left open.

```kotlin
fun main() {
    try {
        val controller = LottoController()
        controller.run()
    } catch (e: Exception) {
        OutputView.exit(1)
    }
}
```

Would handling it like this be more appropriate?

---

솔직하게 `Application.kt`에서 이상적으로 어느 정도의 일을 해야하는지 따로 생각해보지 않았어요. 기존에 사용하던 언어에서 `int main(int ac, char** av)` 상의 일 정도만 한다고 생각을 했거든요. 그 부분은 다시 공부해보고 피드백 적용 하겠습니다. 

`view` package 에서 아래 부분을 handling 하는게 더 적절하다고 말씀하신 걸까요?

```kotlin
println(e.message ?: "Unexpected error occurred.")
```

extra safety로, `try/catch`를 하나 더  만들어서 `illegalargumentexception` 외의 `throw`가 던져질 시 terminate으로 처리하자 까지만 합의했었습니다. `OutputView`로 넘길까 하기도 했는데, `println()`가 노출된 상태로 메세지 처리를 계속 할것이냐에 대한 결론을 재대로 내지 않고, 이야기하다가 넘어갔거든요. 

```kotlin
fun main() {
    try {
        val controller = LottoController()
        controller.run()
    } catch (e: Exception) {
       OutputView.exit(1)
    }
}
```

같은 방식으로 처리하는게 더 맞을까요?

File: src/main/kotlin/lotto/controller/LottoController.kt (hunk 4 above)
Comment (miooo0o) [reply src/main/kotlin/lotto/controller/LottoController.kt]: I rushed it without thinking through the class/object design properly. Agree. This class should be object. 
We did talk about it, but I forgot to actually apply it and didn’t update the feature list properly either.  
Since I’ll be refactoring the structure a lot anyway, I’ll note it down now and revisit this part again in step2.

File: src/main/kotlin/lotto/util/LottoUtils.kt (hunk 5 above)
Comment (miooo0o) [reply src/main/kotlin/lotto/util/LottoUtils.kt]: I agree. Extracting this part as `Utils` was honestly an easy choice.
It would have been better to define `Lotto` as a data class and extend it properly, but I still haven’t clearly figured out how much responsibility and functionality a data class should ideally have.
Today, Jason showed an example of what we could do for step2, and I think that gave me a bit of a hint.

```kotlin
data class LottoNumber private constructor(private val value: Int) {
    companion object {
        internal const val MIN_RANGE_NUMBER = 1
        internal const val MAX_RANGE_NUMBER = 45
        private val LOTTO_NUMBERS = (MIN_RANGE_NUMBER..MAX_RANGE_NUMBER).associateWith(::LottoNumber)

        fun from(value: Int): LottoNumber {
            return LOTTO_NUMBERS[value] ?: throw IllegalArgumentException()
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as LottoNumber
        return value == other.value
    }

    override fun hashCode(): Int {
        return value
    }
}

```
I gonna use this for step2.
Thanks a lot for the helpful feedback!

---

맞아요. 동의합니다. 이 부분을 `Utils`로 빼는건 쉬운 선택이였어요.
`Lotto` class를 data class로 정의하고 확장했어야했다고 생각하는데, 저는 아직도 `data class`같은 새로운 타입이 얼만큼의 기능을 가지는게 옳은지에 대해, 재대로 확립하지 못한 것 같아요.

오늘 Jason이 step2를 위해 무엇을 할 수 있을지 예시를 보여주었는데 그게 조금의 힌트가 될 듯 합니다...!

좋은 피드백 감사드립니다.

File: src/main/kotlin/lotto/view/InputView.kt (hunk 16 above)
Comment (miooo0o) [reply src/main/kotlin/lotto/view/InputView.kt]: This is actually something Sebas and I  talked about once before, but writing this comment made me realize that we did tend to stick with the ‘easy choice’ even when refactoring.  I’ll try to organize things more clearly in step2.  

--- 
이부분도 한번 이야기 했던 건데, comment를 달다 보니 저희가 '쉬운 선택'을 한 부분은 확실히 refactor시에도 그대로 유지한 경향이 있었다는게 느껴지네요. step2에서는 더 잘 분류헤보겠습니다.

File: src/main/kotlin/lotto/services/LottoService.kt (hunk 19 above)
Comment (miooo0o) [reply src/main/kotlin/lotto/services/LottoService.kt]: I agree. This was also one of the biggest “easy choices”, Sebas and I talked about but never update. Honestly, it was the part that bothered me the most, so I have nothing to say in my defense 😭… You’re right. I’ll make sure to split everything properly this time.

But what do you think should be the criteria for separating and defining things as `domain`?

File: src/test/kotlin/lotto/domain/LottoTest.kt (hunk 18 above)
Comment (miooo0o) [reply src/test/kotlin/lotto/domain/LottoTest.kt]: Good point. I guess I was only checking for extreme cases, I'll reflect!
Thanks.

File: src/main/kotlin/lotto/util/LottoUtils.kt (hunk 5 above)
Comment (miooo0o) [reply src/main/kotlin/lotto/util/LottoUtils.kt]: Oh, but if that’s the case, what do you think about overriding `toString` in the `Lotto` class?

```kotlin
    override fun toString(): String {
        return numbers.joinToString(", ", "[", "]")
    }
```

If the class is already responsible for validating itself, then I think having a simple `toString` for basic formatting wouldn’t be too much, even though formatting is technically a different concern.

I do agree that, ideally, presentation should stay in the `view` package, but I feel it’s still reasonable for `Lotto` to have a minimal `toString` for convenience.

**Comment by miooo0o** [comment]
Hi, Hyeon-Gu Choi,

Unfortunately, I misunderstood today’s program. I tried to work on step 2 and reflect your feedback before properly cleaning up my code in step 1, but that didn’t really work out.

Because of that, my `LottoController` still has its main logic mixed in — it only shows the process, but that’s it.

I still have many things I’m missing, such as:

- I’m still not clear on what exactly service, domain, and application should be, so my structure isn’t well organized yet.

- I’m not sure how to wrap things properly:
  like `IssuedTickets`, `LottoTicket`, and `LottoNumber`. Especially, I’m not confident about how to handle raw data and how to adapt it for tests.

- I don’t fully understand Set, HashSet, or LinkedSet:
  I know the names and roughly what they’re for, but I haven’t learned data structures properly.
  So I felt like I just used them without taking enough care to handle them properly. I know my understanding is still shallow, like I pretended to know more than I did, but I still tried to use new features as much as possible.

and I added many questions before started to write code, some of them perhaps not really related with core topic. I am so sorry....!

thank you for in advance!

File: README.md [hunk 20]
```diff
@@ -1 +1,119 @@
...
+
+---
+
+## [step 2] Features
+
+### Feedback
+
+#### Application
+- [x] separate `println()` into `OutputView`
+- [ ] think/search about what `application` means, and refactor if needed
```
Comment (Hyeon9mak) [inline README.md:70]: > Would handling it like this be more appropriate?

Yep!

And now, when switching from console screen to web browser,
it seems only need to change view package codes!

I think it's more appropriate. How about u?

File: src/main/kotlin/lotto/controller/LottoController.kt [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+
+import lotto.domain.*
+import lotto.services.TicketIssuer
+import lotto.services.TicketsEvaluator
+import lotto.view.InputView
+import lotto.view.OutputView
+
+object LottoController {
+    fun run() {
+        // get amount
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/controller/LottoController.kt:12]: > I’ll note it down now and revisit this part again in step2.

But u changed it well 😄 👍

File: src/main/kotlin/lotto/controller/LottoController.kt [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+        )
+    }
+
+    private fun askBonusNumber(winningInput: Set<Int>): LottoNumber {
+        return InputView.retryUntilSuccess(
+            prompt = { OutputView.Prompt.bonusNumber() },
+            read = { InputView.readBonusNumber(winningInput) }
+        )
+    }
+}
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/controller/LottoController.kt:69]: EOF need new line!
https://hyeon9mak.github.io/github-no-newline-at-a-end-of-file/ 😉 
(it's written in Korean, I think it should be readable enough. 😄)

File: src/main/kotlin/lotto/domain/LottoNumber.kt [hunk 23]
```diff
@@ -0,0 +1,29 @@
+package lotto.domain
+
+data class LottoNumber private constructor(private val value: Int) {
+    companion object {
+        internal const val MIN_RANGE_NUMBER = 1
+        internal const val MAX_RANGE_NUMBER = 45
+        private val LOTTO_NUMBERS = (MIN_RANGE_NUMBER..MAX_RANGE_NUMBER).associateWith(::LottoNumber)
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/domain/LottoNumber.kt:7]: what's difference with non-`internal`?

Comment (Hyeon9mak) [reply src/main/kotlin/lotto/domain/LottoNumber.kt:7]: 1-45 numbers caching 👍 💯

File: src/main/kotlin/lotto/domain/LottoNumber.kt [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+        return value == other.value
+    }
+
+    override fun hashCode(): Int {
+        return value
+    }
+
+    override fun toString(): String {
+        return value.toString()
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/domain/LottoNumber.kt:28]: what is the `data class`'s characteristic? 🤔 
https://kotlinlang.org/docs/data-classes.html

and what is the `value class`'s?
https://kotlinlang.org/docs/inline-classes.html

File: src/main/kotlin/lotto/domain/LottoTicket.kt [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+data class LottoTicket(private val lottoNumbers: HashSet<LottoNumber>) {
+    init {
+        require(lottoNumbers.size == SUFFICIENT_SIZE)
+    }
+
+    fun get() = lottoNumbers
+
+    override fun toString(): String {
+        return lottoNumbers.joinToString(", ")
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/domain/LottoTicket.kt:12]: > Oh, but if that’s the case, what do you think about overriding toString in the Lotto class?

(It's not a definitive answer)
Generally, the `toString()` function (method) is used to indicate 'how to logging for engineer'.
In the examples mentioned earlier -- where the output format changes from console text-based to web browser images -- you can see that `toString()` cannot be meaningfully used. 😃

File: src/main/kotlin/lotto/services/TicketsEvaluator.kt [hunk 26]
```diff
@@ -0,0 +1,18 @@
+package lotto.services
+
+import lotto.domain.Rank
+import lotto.domain.WinningCombination
+import lotto.dto.IssuedTickets
+
+object TicketsEvaluator {
+    fun runWith(issuedTickets: IssuedTickets, winningCombination: WinningCombination): List<Rank> {
+        return issuedTickets.get().map {
+            it.getRank(winningCombination)
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/services/TicketsEvaluator.kt:10]: why those are in `services` package? what is the `service`?
If there's no clear standard for this, I'd recommend refraining from using the expression 'service'.

> what do you think should be the criteria for separating and defining things as domain?

I think the criteria for classifying as domain is 'areas that need to be frequently modified and changed' when considering the characteristics of the business.

For example, if the lottery prize amounts for 1st-5th place change frequently,
it would be very appropriate to classify the Rank enum class as domain.
However, what if the winning rules for each place from 1st-5th are all unique, and especially if only 1st place is more complex and changes frequently?
It would be easier to maintain by separating 1st place into a separate class file.

In this way, the criteria is to make it 'suitable for our business characteristics', 'in a form that's easy to maintain', and 'so that others can understand it at a glance'.

This is also why engineers need to be interested not only in technology but also in business, and need to have extensive maintenance experience. 😉

File: src/main/kotlin/lotto/exceptions/LottoException.kt [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+        )
+
+    class InvalidWinningNumbersFormatException(input: String) :
+        LottoException("Invalid winning numbers format: '$input'. Expected numbers separated by commas.")
+
+    class InvalidWinningNumbersException() :
+        LottoException(
+            "Invalid winning numbers: Expected ${LottoTicket.SUFFICIENT_SIZE} numbers in range + " +
+                    "${LottoNumber.MIN_RANGE_NUMBER} to ${LottoNumber.MAX_RANGE_NUMBER}\n",
+        )
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/exceptions/LottoException.kt:23]: Are these all in use right now? 🤔 

Plz remove the unused code.
Fellow developer will feel confused!

**Review by Hyeon9mak — APPROVED** [review]
Hi Mina! Nice work the reflect feedbacks!

I tried my best to answer your questions in as much detail as possible.
There's probably a lot of abstract and difficult content.
If you have any additional questions, please feel free to ask anytime.

OOP is not originally intended for 'beautiful design', but rather aims for 'maintainable design'.
Therefore, it's natural that it's difficult to understand unless you've directly experienced situations that are easy/difficult to maintain.

Don't be too hard on yourself and take it slow, experiencing things one by one.

I think you'll be able to sufficiently apply the feedback I've given in step2 as well.
Take your time and enjoy step2!

If you have more questions, plz feel free to leave a comment. 👍

File: src/main/kotlin/lotto/domain/LottoNumber.kt (hunk 23 above)
Comment (miooo0o) [reply src/main/kotlin/lotto/domain/LottoNumber.kt:7]: From what I understand, internal is similar to private, but I can still access it in my tests

File: src/main/kotlin/lotto/domain/LottoNumber.kt (hunk 24 above)
Comment (miooo0o) [reply src/main/kotlin/lotto/domain/LottoNumber.kt:28]: Oh, I just learned about it today! there’s something called a `value class`!
Thank you. I still need to practice using it.

- If class need to have only one parameter, I can use a `value class`.
- If I need to extend it further, it’s better to convert it to a `data class`
- good to use for wrapped type is primitive to avoid overhead

File: src/main/kotlin/lotto/domain/LottoTicket.kt (hunk 25 above)
Comment (miooo0o) [reply src/main/kotlin/lotto/domain/LottoTicket.kt:12]: `toString()` is meant for that purpose! 
After reading your comment, I see your point. it doesn’t really make much sense now.
Thanks for the feedback!

File: src/main/kotlin/lotto/services/TicketsEvaluator.kt (hunk 26 above)
Comment (miooo0o) [reply src/main/kotlin/lotto/services/TicketsEvaluator.kt:10]: Right!
deleting it was definitely the better move. I’ve decided that until I have a solid concept for what a `service` really is, I won’t force myself to use any patterns.

---

I can stick to three simple parts: 
- things that hold logic
- things that combine logic
- pure data or similler

I’ll just separate and name them naturally, without attaching unnecessary labels/package. 

what do you think?

---

Your comment made me aware of something I hadn’t paid enough attention to.
I really appreciate it 👍

File: src/main/kotlin/lotto/exceptions/LottoException.kt (hunk 27 above)
Comment (miooo0o) [reply src/main/kotlin/lotto/exceptions/LottoException.kt:23]: I didn’t even realize I hadn’t used this 🫨 
I’ll remove it in Step 2.

File: README.md (hunk 20 above)
Comment (miooo0o) [reply README.md:70]: Moooore appropriate!
I followed your advice and applied completely separating I/O to Step 2 as well.

It worked much better, was a nice exercise. 
Thanks!

File: src/main/kotlin/lotto/controller/LottoController.kt (hunk 22 above)
Comment (miooo0o) [reply src/main/kotlin/lotto/controller/LottoController.kt:69]: Thank you!
꼼꼼하게 읽어볼게요!
